trie/database.go

```diff
@@ -91,22 +99,33 @@ func prepare(diskdb ethdb.Database, config *Config) *Database {
 	}
 }
 
-// NewDatabase initializes the trie database with default settings, namely
+// NewDatabase initializes the trie database with default settings, note
 // the legacy hash-based scheme is used by default.
-func NewDatabase(diskdb ethdb.Database) *Database {
-	return NewDatabaseWithConfig(diskdb, nil)
-}
-
-// NewDatabaseWithConfig initializes the trie database with provided configs.
-// The path-based scheme is not activated yet, always initialized with legacy
-// hash-based scheme by default.
-func NewDatabaseWithConfig(diskdb ethdb.Database, config *Config) *Database {
-	var cleans int
-	if config != nil && config.Cache != 0 {
-		cleans = config.Cache * 1024 * 1024
+func NewDatabase(diskdb ethdb.Database, config *Config) *Database {
+	// Sanitize the config and use the default one if it's not specified.
+	if config == nil {
+		config = HashDefaults
+	}
+	var preimages *preimageStore
+	if config.Preimages {
+		preimages = newPreimageStore(diskdb)
+	}
+	db := &Database{
+		config:    config,
+		diskdb:    diskdb,
+		preimages: preimages,
+	}
+	// Hash-based backend is preferred in case both modes are configured, or neither
+	// of them are configured. Flip when the default scheme is switched.
+	if config.HashDB != nil && config.PathDB != nil {
+		config.PathDB = nil
+		log.Warn("Both 'hash' and 'path' mode are configured, use 'hash' mode as default")
+	}
```

Comment:
This seems a bit odd. IMO if both are specified, we should just bail out as something seems screwy? Should we also check if the requested one matches the one on disk?

Reply:
I will just log.Crit here, it's never expected to happen.

> Should we also check if the requested one matches the one on disk?

It's already checked in the CLI flag level.